File: LeetCode/1248. Count Number of Nice Subarrays/1248.cs
using System;
using System.Collections.Generic;

public class Solution {
    public int NumberOfSubarrays(int[] nums, int k) {
        // Dictionary to store the count of prefix sums
        var count = new Dictionary<int, int>();
        count[0] = 1; // Initial condition to handle the case when prefixSum == k
        int prefixSum = 0, result = 0;

        foreach (int num in nums) {
            // Increment prefixSum by 1 if the current number is odd
            prefixSum += num % 2;
            
            // Check if there exists a prefix sum that would make the subarray have exactly k odd numbers
            if (count.ContainsKey(prefixSum - k)) {
                result += count[prefixSum - k];
            }
            
            // Update the count of the current prefixSum
            if (!count.ContainsKey(prefixSum)) {
                count[prefixSum] = 0;
            }
            count[prefixSum]++;
        }

        return result; // Return the total number of nice subarrays
    }

    public static void Main(string[] args) {
        var solution = new Solution();
        int[] nums = {1, 1, 2, 1, 1};
        int k = 3;
        Console.WriteLine(solution.NumberOfSubarrays(nums, k)); // Expected output: 2
    }
}
//other people's solutions
/*
public class Solution {
    public int NumberOfSubarrays(int[] nums, int k) {
        var list = new List<int> {1};
        var total = 0;
        foreach (var num in nums) {
            if (num % 2 == 1) {
                list.Add(1);
            } else {
                list[^1]++;
            }

            if (list.Count - k  - 1 >= 0) {
                total += list[list.Count - k - 1];
            }
        }
        return total;
    }
}
*/
/*
public class Solution {
    public int NumberOfSubarrays(int[] nums, int k) {
        List<int> odd=new List<int>();
        for(int i=0;i<nums.Length;i++)
        {
            if(nums[i]%2==1)
            {
                odd.Add(i);
            }
        }
         odd.Add(nums.Length);

         int first=0;
         int total=0;
         for(int i=0;(i+k)<odd.Count;i++)
         {

                if(i==0)
                total+=((odd[i]+1)*(odd[i+k]-odd[i+k-1]));
                else
                total+=((odd[i]-odd[i-1])*(odd[i+k]-odd[i+k-1]));
         }
         return total;
    }
}
*/
/*
public class Solution {
    public int NumberOfSubarrays(int[] nums, int k) {
        return GetSubSetSum(nums, k) - GetSubSetSum(nums, k - 1);
    }
    private static int GetSubSetSum(int[] nums, int goal){
        if(goal < 0){
            return 0;
        }

        int l = 0, sum = 0, count = 0;
        for(int r = 0; r < nums.Length; r++){
            sum += nums[r] % 2 == 1 ? 1 : 0;
            while(sum > goal){
                sum -= nums[l++] % 2 == 1 ? 1 : 0;
            }

            count += (r - l + 1);
        }
        return count;
    }
}
*/

File: LeetCode/1248. Count Number of Nice Subarrays/1248.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int numberOfSubarrays(int[] nums, int k) {
        // Map to store the count of prefix sums
        Map<Integer, Integer> count = new HashMap<>();
        count.put(0, 1); // Initial condition to handle the case when prefixSum == k
        int prefixSum = 0, result = 0;

        for (int num : nums) {
            // Increment prefixSum by 1 if the current number is odd
            prefixSum += num % 2;
            
            // Check if there exists a prefix sum that would make the subarray have exactly k odd numbers
            if (count.containsKey(prefixSum - k)) {
                result += count.get(prefixSum - k);
            }
            
            // Update the count of the current prefixSum
            count.put(prefixSum, count.getOrDefault(prefixSum, 0) + 1);
        }

        return result; // Return the total number of nice subarrays
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        int[] nums = {1, 1, 2, 1, 1};
        int k = 3;
        System.out.println(solution.numberOfSubarrays(nums, k)); // Expected output: 2
    }
}
//other people's solution
/*
 class Solution {
    public int numberOfSubarrays(int[] nums, int k) {
        int n = nums.length;
        int[] cnt = new int[n + 1];
        cnt[0] = 1;
        int ans = 0, t = 0;
        for (int v : nums) {
            t += v & 1;
            if (t - k >= 0) {
                ans += cnt[t - k];
            }
            cnt[t]++;
        }
        return ans;
    }
}
 */
/*
 class Solution {
    public static int numberOfSubarrays(int[] nums, int k) {
        int count = 0;
        int oddCount = 0;
        int[] prefixCounts = new int[nums.length + 1];
        prefixCounts[0] = 1; // Initialize the prefix count for zero odd numbers.

        for (int num : nums) {
            if (num % 2 != 0) {
                oddCount++;
            }
            if (oddCount >= k) {
                count += prefixCounts[oddCount - k];
            }
            prefixCounts[oddCount]++;
        }

        return count;
    }
}
 */
/*
 class Solution {
    public int numberOfSubarrays(int[] nums, int k) {
        // brute force approach
        // generate all subarrays

        return helper(nums, k) - helper(nums, k - 1);
    }

    private int helper(int[] arr, int k){
        if(k < 0){
            return 0;
        }

        int count = 0;
        int cnt = 0;
        int l = 0;
        int r = 0;

        while(r < arr.length){
            if((arr[r] & 1) == 1){
                cnt++;
            }

            while(cnt > k){
                if((arr[l] & 1) == 1){
                    cnt--;
                }
                l++;
            }

            if(cnt <= k){
                count += (r - l + 1);
            }
            r++;
        }
        return count;
    }
}
 */
/*
 class Solution {
    public int numberOfSubarrays(int[] nums, int k) {
        if (k< 0) {
            return 0;
        }

        int res = count(nums, k) - count(nums, k - 1);
        return res;
    }

    public static int count(int[] nums, int k) {
    int l = 0;
    int r = 0;
    int count = 0;
    int sum = 0;

    while (r < nums.length) {
        sum += (nums[r]%2);

        while (sum > k && l <= r) {
            sum -= (nums[l]%2);
            l++;
        }

        count += (r - l + 1);
        r++;
    }

    return count;
    }
}
 */
/*
 class Solution {
    public int numberOfSubarrays(int[] nums, int k) {
        List<Integer> even = new ArrayList<>();
        int temp = 0, ans = 0; 
        for(int num : nums) {
            if(num % 2 == 0)
                temp++;
            else {
                even.add(temp);
                temp = 0; 
            }
        }
        // 1 1 2 1 1  num = 1 even=[0, 0, 1, 0, 0 ] 
        even.add(temp); 
        int m = even.size();
        //i=0 to i<5-3 =2 
        for(int i = 0; i < m - k; i++) {
            //i=0 ans=ans+(0+1)*(0+1) =0+1=1
            //i=1 ans= 1+(0+1)*(0+1)=2
            
            ans += (even.get(i) + 1) * (even.get(i + k) + 1);
        }
        return ans;
    }
}
 */